emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java

```diff
@@ -0,0 +1,277 @@
+                    .filter(oldViolation -> hasSameLineNumber(oldViolation.get(1), Integer.parseInt(oldViolation.get(2)),
+                            Integer.parseInt(newViolation.get(2)))) // same line number
+                    .collect(Collectors.toSet());
+
+            if (!relevantOldViolations.isEmpty()) {
+                violationsToRemove.add(newViolation);
+            }
+        }
+        for (List<String> violationToRemove : violationsToRemove) {
+            newViolations.remove(violationToRemove);
+        }
+    }
+
+    /**
+     * Determines whether an old class has been renamed to the new one or not.
+     *
+     * @param oldClass Previous possible name of a class
+     * @param newClass Rename being considered
+     * @return Whether the old class name was renamed to the new one
+     */
+    private boolean isRenamed(String oldClass, String newClass) {
+        for (String renamedClass : renames.keySet()) {
+            if (renamedClass.contains(newClass) && renames.get(renamedClass).contains(oldClass)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether an old line in a file can be mapped to the new line. There's room for optimization.
+     *
+     * @param classInfo Particular class being considered (if the class was renamed, this is the old name)
+     * @param oldLine Original line number
+     * @param newLine New line number
+     * @return Whether the original line number can be mapped to the new line number in the updated version
+     */
+    private boolean hasSameLineNumber(String classInfo, int oldLine, int newLine) {
+        for (String className : lineChanges.keySet()) {
+            if (className.contains(classInfo)) {
+                int offset = 0;
+                for (Integer originalLine : lineChanges.get(className).keySet()) {
+                    if (originalLine <= oldLine) {
+                        offset += lineChanges.get(className).get(originalLine);
+                    }
+                }
+                if (newLine >= oldLine) { // if lines have been inserted
+                    return offset >= 0 && offset >= newLine - oldLine;
+                } else { // if lines have been removed
+                    return offset <= 0 && offset <= newLine - oldLine;
+                }
+            }
+        }
+        return oldLine == newLine;
+    }
+
+    /**
+     * Rewrites violation-counts to only include violations in newViolations.
+     */
+    private void rewriteViolationCounts() throws MojoExecutionException {
+        // for each line of violation-counts, if it can be mapped to a new violation it gets to stay (else it goes)
+        try {
+            Path vc = Paths.get(System.getProperty("user.dir"), "violation-counts");
+            List<String> lines = Files.readAllLines(vc);
+            PrintWriter writer = new PrintWriter(vc.toFile());
+            for (String line : lines) {
```

Comment:
I feel like you can optimize the whole routine by making a `Violation` class, and having a `printViolation()` method there that outputs the usual error message that you'd see in `violation-counts`. That way, you don't have to reread the File, you can just overwrite whatever was there with messages from the set of new specs you have.

Reply:
I think you're right that we can reconstruct the error message from a violation (we would have to remember the count at the beginning as well as the function the violation occurred, along with everything else) if I have extra time this week I can look into optimizing it but I don't think it's high on the priority list - do you think you could maybe raise an issue?

Reply:
Sure thing, I can make an issue on this! This is definitely more of an optimization thing and we should prioritize getting all of the features out there :)